Lesson 3. ActiveRecord query interface

AntonBuyanov opened the pull request:
Create model UserPassedTest. Initial data in seeds.rb Class method model Test. Instance method model User.

Commit:
feat: ActiveRecord query interface

Create model UserPassedTest. Initial data in seeds.rb Class method model Test. Instance method model User.

VidgarVii commented on db/seeds.rb:
```diff
@@ -5,3 +5,48 @@
+                 { body: 'Вариант 1 вопрос 3', question_id: questions[2].id, correct: true },
+                 { body: 'Вариант 2 вопрос 3', question_id: questions[2].id, correct: false }
+               ])
+
```
Очень мало данных. Такие тесты будет не возможно тестировать. Стоит добавить хотя бы по 3-4 вопроса для каждого теста и по 4 ответа к каждому вопросу

VidgarVii commented on db/seeds.rb:
```diff
@@ -5,3 +5,48 @@
+               ])
+
+UserPassedTest.create!([
+                         { user_id: users[0].id, test_id: tests[1].id },
```
а вот UserPassedTest можно было не создавать)

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,2 +1,5 @@
 class Test < ApplicationRecord
+  def self.find_test_by_category(name_category)
+    Test.where(categories: { title: name_category }).order(title: :desc).pluck(:title)
```
можно (и рекомендуется)  писать так:
```ruby
where(...)...
```
т.е. без явного указания объекта у которого вызывается метод
потому что в данном случае объектом является класс `Test` и ему равен `self` внутри класс-метода
по дефолту, при вызове метода без указания объекта он вызывается у `self`, т.е. в текущем варианте как раз у `Test`

Commit:
fix: ActiveRecord query interface

Initial data in seeds.rb. Fixed method model Test. Fixed instance method model User.

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,2 +1,8 @@
 class Test < ApplicationRecord
+  belongs_to :category
```
с ассоциациями ты поторопился, это будет чуть дальше в курсе
в этом задании нужно все решать через SQL, с помощью  ActiveRecord queries, но без использования ассоциаций
суть урока - понять какую работу делает за нас `ActiveRecord` и для этого нужно здесь выполнить ее самому
давай сделаем следующее:
- избавься от ассоциаций (оставь связи только на уровне БД, но не на уровне моделей)
- перепиши методы в моделях `User`, `Test` без использования ассоциаций

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,2 +1,8 @@
+  def self.find_test_by_category(name_category)
+    joins(:category)
+      .where(categories: { title: name_category })
+  end
```
ТЗ: "массив названий всех Тестов"
а здесь мы получим объекты целиком

VidgarVii replied:
и всё-ровно тут придёт коллекция тестов. А надо массив названий тестов
https://apidock.com/rails/ActiveRecord/Calculations/pluck

VidgarVii replied:
```ruby
  joins('JOIN categories ON categories.id = tests.category_id')
    .where(categories: { title: name_category })
    .order(:title)
    .pluck(:title)
```

VidgarVii commented on app/models/user.rb:
```diff
@@ -1,2 +1,7 @@
+
+  def level_test(level)
+    tests.where(level: level)
+  end
```
тут тоже внимательно прочти задание

VidgarVii replied:
аргумент для where стоит переписать в хеш. Будет на много лучше читаться

Commit:
fix: ActiveRecord query interface

Fix methods User, Test

Commit:
fix: ActiveRecord query interface

Test method outputs an array of test names. Fix method User

VidgarVii commented on db/migrate/20220601145632_create_user_passed_tests.rb:
```diff
@@ -0,0 +1,10 @@
+  def change
+    create_table :user_passed_tests do |t|
+      t.references :user, foreign_key: true
+      t.references :test, foreign_key: true
```
null false

VidgarVii commented on db/migrate/20220601145632_create_user_passed_tests.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateUserPassedTests < ActiveRecord::Migration[6.1]
+  def change
+    create_table :user_passed_tests do |t|
+      t.references :user, foreign_key: true
```
null false

Commit:
fix: ActiveRecord query interface
